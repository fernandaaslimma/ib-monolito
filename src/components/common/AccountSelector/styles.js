import styled from "styled-components";
import { lighestgrey } from "../../../styles/settings";
import { rem } from "../../../styles/tools";

export const RadioWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Separator = styled.span`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: ${rem(1)};
  background: ${lighestgrey};
`;

export const Space = styled.span`
  display: block;
  margin: 0 ${rem(8)};
  position: relative;
`;

export const BtnWrapper = styled.div`
  flex-grow: 1;
  position: relative;
  flex-basis: 0;
`;
