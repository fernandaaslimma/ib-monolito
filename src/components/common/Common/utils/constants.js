export const PT_BR = "pt-BR";
export const EN_US = "en-US";
export const ZH_CN = "zh-CN";
export const SCREEN_LG = 1199;
export const SCREEN_MD = 992;
export const SCREEN_SM = 767;
export const SCREEN_XS = 449;
export const SCREEN_XXS = 319;
export const UP = "up";
export const DOWN = "down";
export const RIGHT = "right";
export const PAGINATION_TOTAL = "X-Total-Count";
export const DEFAULT_MOBILE_HEADER_HEIGHT = 60;
export const ASCENDING = "ASC";
export const DESCENDING = "DESC";
