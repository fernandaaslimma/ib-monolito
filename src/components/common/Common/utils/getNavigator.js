export const isMsBrowser = () => {
  if (typeof window !== "undefined") {
    return !!(navigator && navigator.msSaveOrOpenBlob);
  }
};

export const isMsEdge = () => {
  if (typeof window !== "undefined") {
    return !!(
      navigator &&
      navigator.msSaveOrOpenBlob &&
      navigator.appVersion.includes("Edge")
    );
  }
};

export const isInternetExplorer = () => isMsBrowser() && !isMsEdge();
