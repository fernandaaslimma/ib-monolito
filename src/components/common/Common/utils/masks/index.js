export const cpfMask = () => {
  return [
    /\d/,
    /\d/,
    /\d/,
    ".",
    /\d/,
    /\d/,
    /\d/,
    ".",
    /\d/,
    /\d/,
    /\d/,
    "-",
    /\d/,
    /\d/
  ];
};

export const cnpjMask = () => {
  return [
    /\d/,
    /\d/,
    ".",
    /\d/,
    /\d/,
    /\d/,
    ".",
    /\d/,
    /\d/,
    /\d/,
    "/",
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    "-",
    /\d/,
    /\d/
  ];
};

export const cepMask = () => {
  return [/\d/, /\d/, /\d/, /\d/, /\d/, "-", /\d/, /\d/, /\d/];
};

export const datetimeMask = [
  /\d/,
  /\d/,
  "/",
  /\d/,
  /\d/,
  "/",
  /\d/,
  /\d/,
  /\d/,
  /\d/
];

export const maxValueMonth = [
  31,
  31,
  29,
  31,
  30,
  31,
  30,
  31,
  31,
  30,
  31,
  30,
  31
];

export const telephoneMask = [
  "(",
  /\d|\*/,
  /\d|\*/,
  ")",
  " ",
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  "-",
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/
];

export const cellphoneMask = [
  "(",
  /\d|\*/,
  /\d|\*/,
  ")",
  " ",
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  "-",
  /\d|\*/,
  /\d|\*/,
  /\d|\*/,
  /\d|\*/
];

export function telMask(value) {
  const raw = value.toString().replace(/\D/g, "");
  if (raw.length > 10) {
    return cellphoneMask;
  } else {
    return telephoneMask;
  }
}

export const formatOrder = ["yyyy", "yy", "mm", "dd", "HH", "MM", "SS"];
export const correctedDatePipe = (
  dateFormat,
  { minYear = 1, maxYear = 9999 } = {}
) => {
  const dateFormatArray = dateFormat
    .split(/[^dmyHMS]+/)
    .sort((a, b) => formatOrder.indexOf(a) - formatOrder.indexOf(b));
  return function(conformedValue) {
    const indexesOfPipedChars = [];
    const maxValue = {
      dd: 31,
      mm: 12,
      yy: 99,
      yyyy: maxYear,
      HH: 23,
      MM: 59,
      SS: 59
    };
    const minValue = {
      dd: 1,
      mm: 1,
      yy: 0,
      yyyy: minYear,
      HH: 0,
      MM: 0,
      SS: 0
    };
    const conformedValueArr = conformedValue.split("");

    // Check first digit
    dateFormatArray.forEach(format => {
      const position = dateFormat.indexOf(format);
      const maxFirstDigit = parseInt(
        maxValue[format].toString().substr(0, 1),
        10
      );

      if (parseInt(conformedValueArr[position], 10) > maxFirstDigit) {
        conformedValueArr[position + 1] = conformedValueArr[position];
        conformedValueArr[position] = 0;
        indexesOfPipedChars.push(position);
      }
    });

    // Check for invalid date
    let month = 0;
    const isInvalid = dateFormatArray.some(format => {
      const position = dateFormat.indexOf(format);
      const length = format.length;
      const textValue = conformedValue
        .substr(position, length)
        .replace(/\D/g, "");
      const value = parseInt(textValue, 10);
      if (format === "mm") {
        month = value || 0;
      }
      const maxValueForFormat =
        format === "dd" ? maxValueMonth[month] : maxValue[format];
      if (format === "yyyy" && (minYear !== 1 || maxYear !== 9999)) {
        const scopedMaxValue = parseInt(
          maxValue[format].toString().substring(0, textValue.length),
          10
        );
        const scopedMinValue = parseInt(
          minValue[format].toString().substring(0, textValue.length),
          10
        );
        return value < scopedMinValue || value > scopedMaxValue;
      }
      return (
        value > maxValueForFormat ||
        (textValue.length === length && value < minValue[format])
      );
    });

    if (isInvalid) {
      return false;
    }

    return {
      value: conformedValueArr.join(""),
      indexesOfPipedChars
    };
  };
};
