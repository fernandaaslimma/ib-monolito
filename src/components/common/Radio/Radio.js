import React from "react";
import { bool, string, func, oneOfType, arrayOf } from "prop-types";
import { RadioTag, Check, Label, SubLabel, UpperLabel } from "./styles";
import { v4 as uuidv4 } from "uuid";

function Radio({
  label,
  upperLabel,
  subLabels,
  onChange,
  checked,
  name,
  disabled,
  value,
  id = uuidv4(),
  dataTest,
  dataTestLabel
}) {
  return (
    <RadioTag
      htmlFor={id}
      disabled={disabled}
      data-test={dataTestLabel}
      hasSub={subLabels && subLabels.length}
    >
      <input
        type="radio"
        id={id}
        checked={checked}
        value={value}
        name={name}
        disabled={disabled}
        onChange={e => onChange(e)}
        data-test={dataTest}
      />
      <Check />
      <Label>
        {upperLabel && <UpperLabel>{upperLabel}</UpperLabel>}
        {label}
        {subLabels &&
          subLabels.length > 0 &&
          subLabels.map((subLabel, index) => {
            return <SubLabel key={index}>{subLabel}</SubLabel>;
          })}
      </Label>
    </RadioTag>
  );
}

Radio.displayName = "Radio";

Radio.defaultProps = {
  dataTestLabel: "Radio",
  dataTest: "RadioInput"
};

Radio.propTypes = {
  label: string,
  upperLabel: string,
  subLabels: arrayOf(string),
  onChange: func,
  checked: bool,
  disabled: bool,
  name: string,
  value: oneOfType([bool, string])
};

export default Radio;
