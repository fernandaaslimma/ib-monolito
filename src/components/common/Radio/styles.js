import styled, { css } from "styled-components";
import { rem, remFontSize } from "../../../styles/tools";
import { white, darkBlue, grey70, gray200 } from "../../../styles/settings";

export const Check = styled.div`
  border: solid ${grey70} ${rem(2)};
  border-radius: 100%;
  width: ${rem(14)};
  height: ${rem(14)};
  position: relative;
  overflow: hidden;
  transition: all 0.2s linear;

  :before {
    content: "";
    display: block;
    width: 100%;
    height: 100%;
    box-sizing: border-box;
    border-radius: 100%;
    background: transparent;
    border: solid ${white} ${rem(2)};
  }
`;

export const RadioTag = styled.label`
  display: inline-flex;
  align-items: center;
  border-radius: ${rem(4)};
  cursor: pointer;
  margin: ${rem(10)};
  font-size: ${remFontSize(14)};

  input {
    position: absolute;
    visibility: hidden;
    pointer-events: none;

    :checked ~ ${Check} {
      border: solid ${darkBlue} ${rem(2)};

      :before {
        background: ${darkBlue};
      }
    }
  }

  > span {
    flex-grow: 1;
    margin-left: ${rem(10)};
  }

  ${({ disabled }) =>
    disabled &&
    css`
      input ~ ${Check}, input ~ span {
        opacity: 0.5;
        cursor: default;
    `};

  ${({ hasSub }) =>
    hasSub &&
    css`
      margin: ${rem(16)} ${rem(10)};
    `};
`;

export const Label = styled.span`
  font-size: ${remFontSize(16)};
  font-family: Lato;
`;

export const SubLabel = styled.span`
  font-size: ${remFontSize(14)};
  line-height: ${remFontSize(16)};
  margin-top: ${rem(5)};
  color: ${gray200};
  display: block;
`;

export const UpperLabel = styled.span`
  display: block;
  font-family: Lato;
  font-style: normal;
  font-weight: normal;
  font-size: ${remFontSize(14)};
  line-height: ${remFontSize(16)};
  color: ${gray200};
  margin-bottom: ${rem(4)};
`;
