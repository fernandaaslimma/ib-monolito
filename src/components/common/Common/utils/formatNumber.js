import { EN_US, PT_BR } from "../utils/constants";

export default function formatNumber(
  value,
  numberOptions = {},
  locale = EN_US
) {
  const validLocale =
    [EN_US, PT_BR].find(l => l.toLowerCase() === locale.toLowerCase()) || EN_US;

  const options = {};
  if (numberOptions.currency) {
    options.style = "currency";
    options.currency = numberOptions.currency;
  }

  if (numberOptions.digits) {
    options.minimumFractionDigits = numberOptions.digits;
    options.maximumFractionDigits = numberOptions.digits;
  }

  return new Intl.NumberFormat(validLocale, options).format(value);
}
