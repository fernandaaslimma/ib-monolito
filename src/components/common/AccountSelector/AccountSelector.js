import React, { Fragment, useState } from "react";
import { string, func, array } from "prop-types";
import { RadioWrapper, Space, Separator, BtnWrapper } from "./styles";
import { translate } from "../../../utils/i18n";
import Radio from "../Radio";
import formatNumber, { formatCNPJ } from "../../../utils/formatNumber";
import { BRL_CURRENCY } from "../../../utils/constants";
import { Button } from "react-bocombbm-components";
import { isIndividualUser } from "../../../utils/roles";

function AccountSelector({
  accounts,
  onChange,
  selectedAccountNumber,
  withButton,
  isOpen //only use this prop when the account selector is inside an AnimatedBottomSheet
}) {
  const [changedSelectedAccount, setChangedSelectedAccount] = useState(
    selectedAccountNumber
  );

  //When closing the Sheet, verify the selected account
  if (isOpen === false && selectedAccountNumber !== changedSelectedAccount) {
    setChangedSelectedAccount(selectedAccountNumber);
  }

  // UseEffect for change account outside AnimatedBottomSheet
  React.useEffect(() => {
    if (!isOpen && selectedAccountNumber !== changedSelectedAccount) {
      setChangedSelectedAccount(selectedAccountNumber);
    }
  }, [
    isOpen,
    selectedAccountNumber,
    changedSelectedAccount,
    setChangedSelectedAccount
  ]);

  return (
    <RadioWrapper>
      {accounts &&
        accounts.length &&
        accounts.map((account, index) => {
          const { availableBalance } = account;
          const formatedBalance = formatNumber(availableBalance, {
            digits: 2
          });
          const balanceLabel = translate("EFT_AVAILABLE_BALANCE");

          const subLabels = !isIndividualUser()
            ? [
                `CNPJ: ${formatCNPJ(account.document)}`,
                `${balanceLabel}: ${BRL_CURRENCY}${formatedBalance}`
              ]
            : [`${balanceLabel}: ${BRL_CURRENCY}${formatedBalance}`];

          return (
            <Fragment key={index}>
              <Radio
                dataTestLabel={`Account_${index}`}
                label={account.accountNumber}
                subLabels={subLabels}
                onChange={e =>
                  withButton
                    ? setChangedSelectedAccount(e.target.value)
                    : onChange(e)
                }
                name="number"
                value={account.accountNumber}
                checked={account.accountNumber === changedSelectedAccount}
                disabled={false}
              />
              {account !== accounts[accounts.length - 1] && (
                <Fragment>
                  <Space>
                    <Separator />
                  </Space>
                  <Separator />
                </Fragment>
              )}
            </Fragment>
          );
        })}
      {withButton && accounts && accounts.length > 0 && changedSelectedAccount && (
        <BtnWrapper>
          <Button
            dataTest="saveAccount"
            onClick={() =>
              onChange({ target: { value: changedSelectedAccount } })
            }
            spacing={{ top: "s", bottom: "s", right: "s", left: "s" }}
          >
            {translate("ACCOUNT_SELECTOR_COMPONENT_SAVE_BUTTON")}
          </Button>
        </BtnWrapper>
      )}
    </RadioWrapper>
  );
}

AccountSelector.propTypes = {
  accounts: array.isRequired,
  onChange: func.isRequired,
  selectedAccountNumber: string.isRequired
};

export default AccountSelector;
