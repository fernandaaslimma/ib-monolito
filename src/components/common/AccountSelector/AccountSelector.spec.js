import React from "react";
import { shallow } from "enzyme";
import AccountSelector from "./AccountSelector";
import Radio from "../Radio";
import { Button } from "react-bocombbm-components";

import store from "../../../utils/store";
import { INDIVIDUAL } from "../../../utils/constants";

jest.mock("../../../utils/store", () => ({
  getState: jest.fn()
}));

const props = {
  accounts: [
    { accountNumber: 123, availableBalance: 200.0, document: "11111111111111" },
    { accountNumber: 456, availableBalance: 200.0, document: "11111111111111" }
  ],
  onChange: jest.fn(),
  selectedAccountNumber: 345,
  withButton: false,
  isOpen: false
};

describe("Account Selector", () => {
  let setState, useStateSpy;
  beforeEach(() => {
    setState = jest.fn();
    useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation(init => [init, setState]);

    store.getState.mockImplementation(() => ({
      userInfo: {
        tenants: [INDIVIDUAL]
      }
    }));
  });
  it("should match snapshot closed", () => {
    expect(shallow(<AccountSelector {...props} />)).toMatchSnapshot();
  });

  it("Should change account", () => {
    const shallowComponent = shallow(<AccountSelector {...props} />);

    shallowComponent
      .find(Radio)
      .at(1)
      .prop("onChange")({ target: { value: props.accounts[1] } });

    expect(shallowComponent).toMatchSnapshot();
  });

  it("Should select account with button", () => {
    const newProps = { ...props, withButton: true };
    const shallowComponent = shallow(<AccountSelector {...newProps} />);

    shallowComponent
      .find(Radio)
      .at(1)
      .prop("onChange")({ target: { value: props.accounts[1] } });

    expect(shallowComponent).toMatchSnapshot();
  });

  it("Should select account with button", () => {
    const newProps = { ...props, withButton: true };
    const shallowComponent = shallow(<AccountSelector {...newProps} />);

    shallowComponent
      .find(Radio)
      .at(1)
      .prop("onChange")({ target: { value: props.accounts[1] } });
    shallowComponent.find(Button).simulate("click");
    expect(shallowComponent).toMatchSnapshot();
  });
});
