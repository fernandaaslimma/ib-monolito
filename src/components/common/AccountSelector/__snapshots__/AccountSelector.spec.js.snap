// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Selector Should change account 1`] = `
<styled.div>
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_0"
    disabled={false}
    label={123}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={123}
  />
  <styled.span>
    <styled.span />
  </styled.span>
  <styled.span />
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_1"
    disabled={false}
    label={456}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={456}
  />
</styled.div>
`;

exports[`Account Selector Should select account with button 1`] = `
<styled.div>
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_0"
    disabled={false}
    label={123}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={123}
  />
  <styled.span>
    <styled.span />
  </styled.span>
  <styled.span />
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_1"
    disabled={false}
    label={456}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={456}
  />
  <styled.div>
    <Button
      dataTest="saveAccount"
      disabled={false}
      onClick={[Function]}
      spacing={
        Object {
          "bottom": "s",
          "left": "s",
          "right": "s",
          "top": "s",
        }
      }
    />
  </styled.div>
</styled.div>
`;

exports[`Account Selector Should select account with button 2`] = `
<styled.div>
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_0"
    disabled={false}
    label={123}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={123}
  />
  <styled.span>
    <styled.span />
  </styled.span>
  <styled.span />
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_1"
    disabled={false}
    label={456}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={456}
  />
  <styled.div>
    <Button
      dataTest="saveAccount"
      disabled={false}
      onClick={[Function]}
      spacing={
        Object {
          "bottom": "s",
          "left": "s",
          "right": "s",
          "top": "s",
        }
      }
    />
  </styled.div>
</styled.div>
`;

exports[`Account Selector should match snapshot closed 1`] = `
<styled.div>
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_0"
    disabled={false}
    label={123}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={123}
  />
  <styled.span>
    <styled.span />
  </styled.span>
  <styled.span />
  <Radio
    checked={false}
    dataTest="RadioInput"
    dataTestLabel="Account_1"
    disabled={false}
    label={456}
    name="number"
    onChange={[Function]}
    subLabels={
      Array [
        "undefined: R$200,00",
      ]
    }
    value={456}
  />
</styled.div>
`;
